src/nvim/drawscreen.c

```diff
@@ -2188,18 +2188,11 @@ static void win_update(win_T *wp)
           // rows, and may insert/delete lines
           int j = idx;
           for (l = lnum; l < mod_bot; l++) {
-            linenr_T first = l;
             int prev_rows = new_rows;
-            if (hasFolding(wp, l, NULL, &l)) {
-              new_rows += !decor_conceal_line(wp, first - 1, false);
-            } else if (l == wp->w_topline) {
-              int n = plines_win_nofill(wp, l, false) + wp->w_topfill
-                      - adjust_plines_for_skipcol(wp);
-              n = MIN(n, wp->w_height_inner);
-              new_rows += n;
-            } else {
-              new_rows += plines_win(wp, l, true);
+            if (l == wp->w_topline) {
+              new_rows -= adjust_plines_for_skipcol(wp);
             }
+            new_rows += plines_win_full(wp, l, &l, NULL, true, true);
```

Comment:
`limit_winheight == true` does not look right. With `'smoothscroll'` the return value of `adjust_plines_for_skipcol(wp)` can be large than the window height.